// @burnt-labs/ui/dist/chunk-IHOGI7MH.js
'use strict';

var chunkYHNDTU6H_js = require('./chunk-YHNDTU6H.js');
var chunkPKUHTIDK_js = require('./chunk-PKUHTIDK.js');
var React = require('react');
var PopoverPrimitive = require('@radix-ui/react-popover');
var jsxRuntime = require('react/jsx-runtime');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var PopoverPrimitive__namespace = /*#__PURE__*/_interopNamespace(PopoverPrimitive);

var Popover = PopoverPrimitive__namespace.Root;
var PopoverTrigger = PopoverPrimitive__namespace.Trigger;
var PopoverContent = React__namespace.forwardRef((_a, forwardedRef) => {
  var _b = _a, { className, children } = _b, props = chunkPKUHTIDK_js.__objRest(_b, ["className", "children"]);
  return /* @__PURE__ */ jsxRuntime.jsx(PopoverPrimitive__namespace.Portal, { children: /* @__PURE__ */ jsxRuntime.jsxs(
    PopoverPrimitive__namespace.Content,
    chunkPKUHTIDK_js.__spreadProps(chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn(
        "ui-rounded ui-text-white ui-z-50 ui-p-2 ui-bg-[#434040] ui-data-[state=open]:ui-data-[side=top]:ui-animate-slideDownAndFade ui-data-[state=open]:ui-data-[side=right]:ui-animate-slideLeftAndFade ui-data-[state=open]:ui-data-[side=bottom]:ui-animate-slideUpAndFade ui-data-[state=open]:ui-data-[side=left]:ui-animate-slideRightAndFade",
        className
      ),
      ref: forwardedRef
    }, props), {
      children: [
        children,
        /* @__PURE__ */ jsxRuntime.jsx(PopoverPrimitive__namespace.Arrow, { className: "ui-fill-[#434040]" })
      ]
    })
  ) });
});
PopoverContent.displayName = PopoverPrimitive__namespace.Content.displayName;

exports.Popover = Popover;
exports.PopoverContent = PopoverContent;
exports.PopoverTrigger = PopoverTrigger;
