// @burnt-labs/ui/dist/chunk-KZ3C2OHV.mjs
import { jsxs, jsx } from 'react/jsx-runtime';

// src/icons/keplr.tsx
function KeplrLogo() {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "24",
      height: "24",
      viewBox: "0 0 42 42",
      fill: "none",
      children: [
        /* @__PURE__ */ jsxs("g", { clipPath: "url(#clip0_425_5107)", children: [
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M32.4545 0H9.54545C4.27365 0 0 4.27365 0 9.54545V32.4545C0 37.7264 4.27365 42 9.54545 42H32.4545C37.7264 42 42 37.7264 42 32.4545V9.54545C42 4.27365 37.7264 0 32.4545 0Z",
              fill: "url(#paint0_linear_425_5107)"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M32.4545 0H9.54545C4.27365 0 0 4.27365 0 9.54545V32.4545C0 37.7264 4.27365 42 9.54545 42H32.4545C37.7264 42 42 37.7264 42 32.4545V9.54545C42 4.27365 37.7264 0 32.4545 0Z",
              fill: "url(#paint1_radial_425_5107)"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M32.4545 0H9.54545C4.27365 0 0 4.27365 0 9.54545V32.4545C0 37.7264 4.27365 42 9.54545 42H32.4545C37.7264 42 42 37.7264 42 32.4545V9.54545C42 4.27365 37.7264 0 32.4545 0Z",
              fill: "url(#paint2_radial_425_5107)"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M32.4545 0H9.54545C4.27365 0 0 4.27365 0 9.54545V32.4545C0 37.7264 4.27365 42 9.54545 42H32.4545C37.7264 42 42 37.7264 42 32.4545V9.54545C42 4.27365 37.7264 0 32.4545 0Z",
              fill: "url(#paint3_radial_425_5107)"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M17.2526 32.2614V22.5192L26.7185 32.2614H31.9849V32.0079L21.0964 20.9122L31.1469 10.3857V10.2614H25.8464L17.2526 19.5635V10.2614H12.9849V32.2614H17.2526Z",
              fill: "white"
            }
          )
        ] }),
        /* @__PURE__ */ jsxs("defs", { children: [
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint0_linear_425_5107",
              x1: "21",
              y1: "0",
              x2: "21",
              y2: "42",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "#1FD1FF" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "#1BB8FF" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "radialGradient",
            {
              id: "paint1_radial_425_5107",
              cx: "0",
              cy: "0",
              r: "1",
              gradientUnits: "userSpaceOnUse",
              gradientTransform: "translate(2.00623 40.4086) rotate(-45.1556) scale(67.3547 68.3624)",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "#232DE3" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "#232DE3", stopOpacity: "0" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "radialGradient",
            {
              id: "paint2_radial_425_5107",
              cx: "0",
              cy: "0",
              r: "1",
              gradientUnits: "userSpaceOnUse",
              gradientTransform: "translate(39.7379 41.7602) rotate(-138.45) scale(42.1137 64.2116)",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "#8B4DFF" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "#8B4DFF", stopOpacity: "0" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "radialGradient",
            {
              id: "paint3_radial_425_5107",
              cx: "0",
              cy: "0",
              r: "1",
              gradientUnits: "userSpaceOnUse",
              gradientTransform: "translate(20.6501 0.311498) rotate(90) scale(33.1135 80.3423)",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "#24D5FF" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "#1BB8FF", stopOpacity: "0" })
              ]
            }
          ),
          /* @__PURE__ */ jsx("clipPath", { id: "clip0_425_5107", children: /* @__PURE__ */ jsx("rect", { width: "42", height: "42", fill: "white" }) })
        ] })
      ]
    }
  );
}

export { KeplrLogo };
