// @burnt-labs/ui/dist/chunk-UYDWFDGW.js
'use strict';

var chunkYHNDTU6H_js = require('./chunk-YHNDTU6H.js');
var chunkPKUHTIDK_js = require('./chunk-PKUHTIDK.js');
var React = require('react');
var DialogPrimitive = require('@radix-ui/react-dialog');
var VisuallyHidden = require('@radix-ui/react-visually-hidden');
var jsxRuntime = require('react/jsx-runtime');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var DialogPrimitive__namespace = /*#__PURE__*/_interopNamespace(DialogPrimitive);
var VisuallyHidden__namespace = /*#__PURE__*/_interopNamespace(VisuallyHidden);

var Dialog = DialogPrimitive__namespace.Root;
var DialogTrigger = DialogPrimitive__namespace.Trigger;
var DialogPortal = DialogPrimitive__namespace.Portal;
var DialogClose = DialogPrimitive__namespace.Close;
var DialogOverlay = React__namespace.forwardRef((_a, ref) => {
  var _b = _a, { className } = _b, props = chunkPKUHTIDK_js.__objRest(_b, ["className"]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    DialogPrimitive__namespace.Overlay,
    chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn(
        "ui-fixed ui-inset-0 ui-z-50 ui-backdrop-blur-lg ui-bg-black/80 ui-data-[state=open]:ui-animate-in ui-data-[state=closed]:ui-animate-out ui-data-[state=closed]:ui-fade-out-0 ui-data-[state=open]:ui-fade-in-0",
        className
      ),
      ref
    }, props)
  );
});
DialogOverlay.displayName = DialogPrimitive__namespace.Overlay.displayName;
var DialogContent = React__namespace.forwardRef((_a, ref) => {
  var _b = _a, { className, children } = _b, props = chunkPKUHTIDK_js.__objRest(_b, ["className", "children"]);
  return /* @__PURE__ */ jsxRuntime.jsxs(DialogPortal, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(DialogOverlay, {}),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "ui-absolute ui-h-screen ui-w-screen ui-inset-0 ui-bg-modal-overlay ui-backdrop-blur-3xl ui-opacity-40 ui-bg-no-repeat ui-bg-cover ui-bg-center ui-bg-fixed ui-z-50" }),
    /* @__PURE__ */ jsxRuntime.jsxs(
      DialogPrimitive__namespace.Content,
      chunkPKUHTIDK_js.__spreadProps(chunkPKUHTIDK_js.__spreadValues({
        "aria-describedby": void 0,
        className: chunkYHNDTU6H_js.cn(
          "ui-z-50 ui-fixed ui-grid ui-w-full ui-max-w-lg ui-gap-4 ui-p-10 ui-duration-200 sm:ui-rounded-[48px] ui-left-[50%] ui-top-[50%] sm:ui-bg-black/20 !ui-flex ui-justify-center ui-flex-col sm:ui-block sm:ui-flex-none ui-h-screen sm:ui-h-auto sm:ui-backdrop-blur-2xl ui-data-[state=open]:ui-animate-in ui-data-[state=closed]:ui-animate-out ui-data-[state=closed]:ui-fade-out-0 ui-data-[state=open]:ui-fade-in-0 ui-data-[state=closed]:ui-zoom-out-95 ui-data-[state=open]:ui-zoom-in-95 ui-data-[state=closed]:ui-slide-out-to-left-1/2 ui-data-[state=closed]:ui-slide-out-to-top-[48%] ui-data-[state=open]:ui-slide-in-from-left-1/2 ui-data-[state=open]:ui-slide-in-from-top-[48%] ui-translate-x-[-50%] ui-translate-y-[-50%]",
          className
        ),
        ref
      }, props), {
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(VisuallyHidden__namespace.Root, { children: /* @__PURE__ */ jsxRuntime.jsx(DialogPrimitive__namespace.Title, {}) }),
          children
        ]
      })
    )
  ] });
});
DialogContent.displayName = DialogPrimitive__namespace.Content.displayName;
function DialogHeader(_a) {
  var _b = _a, {
    className
  } = _b, props = chunkPKUHTIDK_js.__objRest(_b, [
    "className"
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn(
        "ui-flex ui-flex-col ui-space-y-1.5 ui-text-center sm:ui-text-left",
        className
      )
    }, props)
  );
}
DialogHeader.displayName = "DialogHeader";
function DialogFooter(_a) {
  var _b = _a, {
    className
  } = _b, props = chunkPKUHTIDK_js.__objRest(_b, [
    "className"
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn(
        "ui-flex ui-flex-col-reverse sm:ui-flex-row sm:ui-justify-end sm:ui-space-x-2",
        className
      )
    }, props)
  );
}
DialogFooter.displayName = "DialogFooter";
var DialogTitle = React__namespace.forwardRef((_a, ref) => {
  var _b = _a, { className } = _b, props = chunkPKUHTIDK_js.__objRest(_b, ["className"]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    DialogPrimitive__namespace.Title,
    chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn(
        "ui-text-lg ui-font-semibold ui-leading-none ui-tracking-tight",
        className
      ),
      ref
    }, props)
  );
});
DialogTitle.displayName = DialogPrimitive__namespace.Title.displayName;
var DialogDescription = React__namespace.forwardRef((_a, ref) => {
  var _b = _a, { className } = _b, props = chunkPKUHTIDK_js.__objRest(_b, ["className"]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    DialogPrimitive__namespace.Description,
    chunkPKUHTIDK_js.__spreadValues({
      className: chunkYHNDTU6H_js.cn("ui-text-neutral-500 ui-text-sm", className),
      ref
    }, props)
  );
});
DialogDescription.displayName = DialogPrimitive__namespace.Description.displayName;

exports.Dialog = Dialog;
exports.DialogClose = DialogClose;
exports.DialogContent = DialogContent;
exports.DialogDescription = DialogDescription;
exports.DialogFooter = DialogFooter;
exports.DialogHeader = DialogHeader;
exports.DialogOverlay = DialogOverlay;
exports.DialogPortal = DialogPortal;
exports.DialogTitle = DialogTitle;
exports.DialogTrigger = DialogTrigger;
