// @burnt-labs/ui/dist/chunk-ZVGFBZIM.mjs
import { cn } from './chunk-RFTE7DQK.mjs';
import { __objRest, __spreadProps, __spreadValues } from './chunk-3TCQU24P.mjs';
import * as React from 'react';
import * as PopoverPrimitive from '@radix-ui/react-popover';
import { jsx, jsxs } from 'react/jsx-runtime';

var Popover = PopoverPrimitive.Root;
var PopoverTrigger = PopoverPrimitive.Trigger;
var PopoverContent = React.forwardRef((_a, forwardedRef) => {
  var _b = _a, { className, children } = _b, props = __objRest(_b, ["className", "children"]);
  return /* @__PURE__ */ jsx(PopoverPrimitive.Portal, { children: /* @__PURE__ */ jsxs(
    PopoverPrimitive.Content,
    __spreadProps(__spreadValues({
      className: cn(
        "ui-rounded ui-text-white ui-z-50 ui-p-2 ui-bg-[#434040] ui-data-[state=open]:ui-data-[side=top]:ui-animate-slideDownAndFade ui-data-[state=open]:ui-data-[side=right]:ui-animate-slideLeftAndFade ui-data-[state=open]:ui-data-[side=bottom]:ui-animate-slideUpAndFade ui-data-[state=open]:ui-data-[side=left]:ui-animate-slideRightAndFade",
        className
      ),
      ref: forwardedRef
    }, props), {
      children: [
        children,
        /* @__PURE__ */ jsx(PopoverPrimitive.Arrow, { className: "ui-fill-[#434040]" })
      ]
    })
  ) });
});
PopoverContent.displayName = PopoverPrimitive.Content.displayName;

export { Popover, PopoverContent, PopoverTrigger };
