// @burnt-labs/ui/dist/dialog.js
'use strict';

var chunkUYDWFDGW_js = require('./chunk-UYDWFDGW.js');
require('./chunk-YHNDTU6H.js');
require('./chunk-PKUHTIDK.js');



Object.defineProperty(exports, 'Dialog', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.Dialog; }
});
Object.defineProperty(exports, 'DialogClose', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogClose; }
});
Object.defineProperty(exports, 'DialogContent', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogContent; }
});
Object.defineProperty(exports, 'DialogDescription', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogDescription; }
});
Object.defineProperty(exports, 'DialogFooter', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogFooter; }
});
Object.defineProperty(exports, 'DialogHeader', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogHeader; }
});
Object.defineProperty(exports, 'DialogOverlay', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogOverlay; }
});
Object.defineProperty(exports, 'DialogPortal', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogPortal; }
});
Object.defineProperty(exports, 'DialogTitle', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogTitle; }
});
Object.defineProperty(exports, 'DialogTrigger', {
  enumerable: true,
  get: function () { return chunkUYDWFDGW_js.DialogTrigger; }
});
