// @rimbu/base/dist/cjs/arr.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod2) => __copyProps(__defProp({}, "__esModule", { value: true }), mod2);

// src/arr.mts
var arr_exports = {};
__export(arr_exports, {
  append: () => append,
  concat: () => concat,
  copySparse: () => copySparse,
  forEach: () => forEach,
  init: () => init,
  insert: () => insert,
  last: () => last,
  map: () => map,
  mapSparse: () => mapSparse,
  mod: () => mod,
  prepend: () => prepend,
  reverse: () => reverse,
  reverseMap: () => reverseMap,
  splice: () => splice,
  tail: () => tail,
  update: () => update
});
module.exports = __toCommonJS(arr_exports);
var import_common = require("@rimbu/common");
function append(array, value) {
  const clone = array.slice();
  clone.push(value);
  return clone;
}
function concat(first, second) {
  if (first.length === 0)
    return second;
  if (second.length === 0)
    return first;
  return first.concat(second);
}
function reverse(array, start = 0, end = array.length - 1) {
  const length = end - start + 1;
  const res = [];
  let arrayIndex = start - 1;
  let resIndex = length - 1;
  while (++arrayIndex <= end)
    res[resIndex--] = array[arrayIndex];
  return res;
}
function forEach(array, f, state = (0, import_common.TraverseState)(), reversed = false) {
  if (state.halted)
    return;
  const { halt } = state;
  if (reversed) {
    let i = array.length;
    while (!state.halted && --i >= 0) {
      f(array[i], state.nextIndex(), halt);
    }
  } else {
    const length = array.length;
    let i = -1;
    while (!state.halted && ++i < length) {
      f(array[i], state.nextIndex(), halt);
    }
  }
}
function map(array, f, indexOffset = 0) {
  const result = [];
  let index = indexOffset;
  let i = -1;
  const length = array.length;
  while (++i < length) {
    result[i] = f(array[i], index++);
  }
  return result;
}
function reverseMap(array, f, indexOffset = 0) {
  const result = [];
  let index = indexOffset;
  let arrayIndex = array.length;
  let resultIndex = 0;
  while (--arrayIndex >= 0)
    result[resultIndex++] = f(array[arrayIndex], index++);
  return result;
}
function prepend(array, value) {
  const clone = array.slice();
  clone.unshift(value);
  return clone;
}
function last(arr) {
  return arr[arr.length - 1];
}
function update(arr, index, updater) {
  if (index < 0 || index >= arr.length)
    return arr;
  const curValue = arr[index];
  const newValue = (0, import_common.Update)(curValue, updater);
  if (Object.is(newValue, curValue))
    return arr;
  const newArr = arr.slice();
  newArr[index] = newValue;
  return newArr;
}
function mod(arr, index, f) {
  if (index < 0 || index >= arr.length)
    return arr;
  const curValue = arr[index];
  const newValue = f(curValue);
  if (Object.is(newValue, curValue))
    return arr;
  const newArr = arr.slice();
  newArr[index] = newValue;
  return newArr;
}
function insert(arr, index, value) {
  const clone = arr.slice();
  clone.splice(index, 0, value);
  return clone;
}
function tail(arr) {
  return arr.slice(1);
}
function init(arr) {
  return arr.slice(0, arr.length - 1);
}
function splice(arr, start, deleteCount, ...items) {
  const clone = arr.slice();
  clone.splice(start, deleteCount, ...items);
  return clone;
}
function copySparse(arr) {
  const clone = [];
  for (const key in arr) {
    clone[key] = arr[key];
  }
  return clone;
}
function mapSparse(arr, f) {
  const result = Array(arr.length);
  for (const key in arr) {
    result[key] = f(arr[key], key);
  }
  return result;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  append,
  concat,
  copySparse,
  forEach,
  init,
  insert,
  last,
  map,
  mapSparse,
  mod,
  prepend,
  reverse,
  reverseMap,
  splice,
  tail,
  update
});
