// @rimbu/common/dist/cjs/traverse-state.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/traverse-state.mts
var traverse_state_exports = {};
__export(traverse_state_exports, {
  TraverseState: () => TraverseState
});
module.exports = __toCommonJS(traverse_state_exports);
var TraverseStateImpl = class {
  constructor(startIndex) {
    this.startIndex = startIndex;
    this.halted = false;
    this.halt = () => {
      this.halted = true;
    };
    this.currentIndex = startIndex;
  }
  nextIndex() {
    return this.currentIndex++;
  }
  reset() {
    this.currentIndex = this.startIndex;
    this.halted = false;
  }
};
function TraverseState(startIndex = 0) {
  return new TraverseStateImpl(startIndex);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TraverseState
});
