// @abstract-money/core/dist/actions/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



























































var _chunkCXRY6HLTcjs = require('../chunk-CXRY6HLT.cjs');
require('../chunk-YMNYIN57.cjs');

// src/actions/public/get-ica-client-address-from-registry.ts
async function getIcaClientAddressFromRegistry({
  cosmWasmClient,
  registryAddress,
  version
}) {
  return _chunkCXRY6HLTcjs.getAbstractModuleAddressFromRegistry.call(void 0, {
    moduleName: "ica-client" /* ICA_CLIENT */,
    cosmWasmClient,
    registryAddress,
    version
  });
}

// src/actions/public/get-ica-client-query-client-from-registry.ts
async function getIcaClientQueryClientFromRegistry({
  cosmWasmClient,
  registryAddress,
  version
}) {
  const icaClientAddress = await getIcaClientAddressFromRegistry({
    cosmWasmClient,
    registryAddress,
    version
  });
  return _chunkCXRY6HLTcjs.getIcaClientQueryClient.call(void 0, { cosmWasmClient, icaClientAddress });
}

// src/actions/public/get-module-factory-version.ts
async function getModuleFactoryVersion({
  cosmWasmClient,
  apiUrl,
  version
}) {
  return _chunkCXRY6HLTcjs.getAbstractModuleVersion.call(void 0, {
    moduleName: "module-factory" /* MODULE_FACTORY */,
    cosmWasmClient,
    apiUrl,
    version
  });
}































































exports.CommonModuleNames = _chunkCXRY6HLTcjs.CommonModuleNames; exports.claimNamespace = _chunkCXRY6HLTcjs.claimNamespace; exports.createAccount = _chunkCXRY6HLTcjs.createAccount; exports.createAccountMonarchy = _chunkCXRY6HLTcjs.createAccountMonarchy; exports.createSubAccount = _chunkCXRY6HLTcjs.createSubAccount; exports.deposit = _chunkCXRY6HLTcjs.deposit; exports.execute = _chunkCXRY6HLTcjs.execute; exports.getAbstractModuleAddressFromRegistry = _chunkCXRY6HLTcjs.getAbstractModuleAddressFromRegistry; exports.getAbstractModuleVersion = _chunkCXRY6HLTcjs.getAbstractModuleVersion; exports.getAccountAddressFromApi = _chunkCXRY6HLTcjs.getAccountAddressFromApi; exports.getAccountBalancesFromApi = _chunkCXRY6HLTcjs.getAccountBalancesFromApi; exports.getAccountClient = _chunkCXRY6HLTcjs.getAccountClient; exports.getAccountClientFromApi = _chunkCXRY6HLTcjs.getAccountClientFromApi; exports.getAccountQueryClient = _chunkCXRY6HLTcjs.getAccountQueryClient; exports.getAccountQueryClientFromApi = _chunkCXRY6HLTcjs.getAccountQueryClientFromApi; exports.getAccountsByOwnerFromApi = _chunkCXRY6HLTcjs.getAccountsByOwnerFromApi; exports.getAnsHostAddressFromApi = _chunkCXRY6HLTcjs.getAnsHostAddressFromApi; exports.getAnsHostAddressFromRegistry = _chunkCXRY6HLTcjs.getAnsHostAddressFromRegistry; exports.getAnsHostClient = _chunkCXRY6HLTcjs.getAnsHostClient; exports.getAnsHostClientFromApi = _chunkCXRY6HLTcjs.getAnsHostClientFromApi; exports.getAnsHostClientFromRegistry = _chunkCXRY6HLTcjs.getAnsHostClientFromRegistry; exports.getAnsHostQueryClient = _chunkCXRY6HLTcjs.getAnsHostQueryClient; exports.getAnsHostQueryClientFromApi = _chunkCXRY6HLTcjs.getAnsHostQueryClientFromApi; exports.getAnsHostQueryClientFromRegistry = _chunkCXRY6HLTcjs.getAnsHostQueryClientFromRegistry; exports.getAnsHostVersionFromApi = _chunkCXRY6HLTcjs.getAnsHostVersionFromApi; exports.getAnsTokenFromApi = _chunkCXRY6HLTcjs.getAnsTokenFromApi; exports.getAnsTokensFromApi = _chunkCXRY6HLTcjs.getAnsTokensFromApi; exports.getAppModuleCodeIdFromRegistry = _chunkCXRY6HLTcjs.getAppModuleCodeIdFromRegistry; exports.getChainRpcUrlFromApi = _chunkCXRY6HLTcjs.getChainRpcUrlFromApi; exports.getCosmWasmClient = _chunkCXRY6HLTcjs.getCosmWasmClient; exports.getIbcClientAddressFromRegistry = _chunkCXRY6HLTcjs.getIbcClientAddressFromRegistry; exports.getIbcClientQueryClient = _chunkCXRY6HLTcjs.getIbcClientQueryClient; exports.getIbcClientQueryClientFromAccount = _chunkCXRY6HLTcjs.getIbcClientQueryClientFromAccount; exports.getIbcClientQueryClientFromRegistry = _chunkCXRY6HLTcjs.getIbcClientQueryClientFromRegistry; exports.getIcaClientAddressFromRegistry = getIcaClientAddressFromRegistry; exports.getIcaClientQueryClient = _chunkCXRY6HLTcjs.getIcaClientQueryClient; exports.getIcaClientQueryClientFromRegistry = getIcaClientQueryClientFromRegistry; exports.getModuleAddress = _chunkCXRY6HLTcjs.getModuleAddress; exports.getModuleFactoryAddressFromRegistry = _chunkCXRY6HLTcjs.getModuleFactoryAddressFromRegistry; exports.getModuleFactoryVersion = getModuleFactoryVersion; exports.getModuleInstantiate2AddressFromApi = _chunkCXRY6HLTcjs.getModuleInstantiate2AddressFromApi; exports.getModules = _chunkCXRY6HLTcjs.getModules; exports.getModulesFromApi = _chunkCXRY6HLTcjs.getModulesFromApi; exports.getNamespace = _chunkCXRY6HLTcjs.getNamespace; exports.getOwner = _chunkCXRY6HLTcjs.getOwner; exports.getRegistryAddressFromApi = _chunkCXRY6HLTcjs.getRegistryAddressFromApi; exports.getRegistryClient = _chunkCXRY6HLTcjs.getRegistryClient; exports.getRegistryClientFromApi = _chunkCXRY6HLTcjs.getRegistryClientFromApi; exports.getRegistryModuleData = _chunkCXRY6HLTcjs.getRegistryModuleData; exports.getRegistryQueryClient = _chunkCXRY6HLTcjs.getRegistryQueryClient; exports.getRegistryQueryClientFromApi = _chunkCXRY6HLTcjs.getRegistryQueryClientFromApi; exports.getRemoteAccountIds = _chunkCXRY6HLTcjs.getRemoteAccountIds; exports.getRemoteAccounts = _chunkCXRY6HLTcjs.getRemoteAccounts; exports.getSenderAddress = _chunkCXRY6HLTcjs.getSenderAddress; exports.getSigningCosmWasmClient = _chunkCXRY6HLTcjs.getSigningCosmWasmClient; exports.getSubAccountIds = _chunkCXRY6HLTcjs.getSubAccountIds; exports.getSubAccountIdsByModuleIdFromApi = _chunkCXRY6HLTcjs.getSubAccountIdsByModuleIdFromApi; exports.getSubAccountSequences = _chunkCXRY6HLTcjs.getSubAccountSequences; exports.revokeNamespace = _chunkCXRY6HLTcjs.revokeNamespace; exports.sendFunds = _chunkCXRY6HLTcjs.sendFunds; exports.upgradeModules = _chunkCXRY6HLTcjs.upgradeModules; exports.withdraw = _chunkCXRY6HLTcjs.withdraw;
